ports/tcpport.js

```diff
@@ -240,7 +242,15 @@ class TcpPort extends EventEmitter {
         });
 
         // send buffer to slave
-        this._client.write(buffer);
+        this._writeCompleted = new Promise((resolve, reject) => {
+            this._writeCompleted.finally(() => {
+                if (this._client.write(buffer)) {
+                    resolve();
+                } else {
+                    this._client.once('drain', resolve);
+                }
+            });
+        });
```

Comment:
```suggestion
       this._writeCompleted = new Promise((resolve, reject) => {
            if (this._client.write(buffer)) {
                resolve();
            } else {
                this._client.once('drain', resolve);
            }
        });

        // Wait for the write operation to complete
        try {
            await this._writeCompleted;
        } catch (error) {
           modbusSerialDebug('An error occurred during the write operation:', error);
            throw error;
        }
```

Reply:
Unfortunately, `write` is not `async`. so we can't await in it. It is my understanding that `_client.write` will emit an `error` event on error instead of throwing, although I can't find any documentation about that online.

Reply:
I think there might be a misconception here:

The `this._writeCompleted.finally` is not waiting on itself, it's waiting on the previous promise that was stored in `this._writeCompleted` and then overwrites `this._writeCompleted` with the new promise. We need to have a Promise because we need a way to signal that we are done from the callback passed to write, which can be called asynchronously.

The only thing I can think of is do the assignment on another line, do you think this would help?
```js
let newWritePromise = this._writeCompleted = new Promise((resolve) => {
    this._writeCompleted.finally(/* ... */);
};
this._writeCompleted = newWritePromise;
```

Reply:
i have some concerns with the new suggestion ( valid to current suggestion too )

`this._writeCompleted` is overwritten each time a new write is initiated. This means if multiple writes are attempted quickly in succession, each write would create a new promise, but the old promise (from a previous write) could be lost. If the older promise hasn't resolved yet (due to waiting for the drain event), it could lead to race conditions where multiple writes might overlap.

If there’s an error in the write process (such as a disconnection), the promise will never resolve or reject, potentially leading to the system hanging indefinitely. For example, if you're managing network connections, where interruptions are common.

can you address the concerns, and since this flow is not clear, can you add comments helping to understand the flow, e.g. what is the current promise we handle, when and why we replace it etc...

Reply:
> but the old promise (from a previous write) could be lost.

I think this is not possible. The function passed as an argument to the `Promise` constructor is executed synchronously before the constructor returns. And as far as I understand Javascript is single threaded, so there is no chance of a race condition where two `TcpPort.write` calls happen simultaneously.

> If the older promise hasn't resolved yet (due to waiting for the drain event), it could lead to race conditions where multiple writes might overlap.

This should also not be possible, because we only call `Socket.write` in the `finally` call of the previous Promise, so there should be no overlap.

> If there’s an error in the write process (such as a disconnection), the promise will never resolve or reject

It is my understanding that `Socket.write` will never throw, but just trigger an `error` event and then return. But I will add a `try catch` to make sure the promise is resolved.

> can you add comments helping to understand the flow

Will do.